src/sqlfluff/rules/L054.py

```diff
@@ -0,0 +1,111 @@
+"""Implementation of Rule L054."""
+from typing import Optional
+
+from sqlfluff.core.rules.base import BaseRule, LintResult, RuleContext
+
+
+class Rule_L054(BaseRule):
+    """Inconsistent column references in GROUP BY/ORDER BY clauses.
+
+    | **Anti-pattern**
+    | A mix of implicit and explicit column references are used in a GROUP BY clause.
+
+    .. code-block:: sql
+       :force:
+
+        SELECT
+            foo,
+            bar,
+            sum(baz) AS sum_value
+        FROM fake_table
+        GROUP BY
+            foo, 2;
+
+        -- The same also applies to column
+        -- references in ORDER BY clauses.
+
+        SELECT
+            foo,
+            bar
+        FROM fake_table
+        ORDER BY
+            1, bar;
+
+    | **Best practice**
+    | Reference all GROUP BY/ORDER BY columns either by name or by position.
+
+    .. code-block:: sql
+       :force:
+
+        -- GROUP BY: Explicit
+        SELECT
+            foo,
+            bar,
+            sum(baz) AS sum_value
+        FROM fake_table
+        GROUP BY
+            foo, bar;
+
+        -- ORDER BY: Explicit
+        SELECT
+            foo,
+            bar
+        FROM fake_table
+        ORDER BY
+            foo, bar;
+
+        -- GROUP BY: Implicit
+        SELECT
+            foo,
+            bar,
+            sum(baz) AS sum_value
+        FROM fake_table
+        GROUP BY
+            1, 2;
+
+        -- ORDER BY: Implicit
+        SELECT
+            foo,
+            bar
+        FROM fake_table
+        ORDER BY
+            1, 2;
+    """
+
+    config_keywords = ["group_by_and_order_by_style"]
+
+    def _eval(self, context: RuleContext) -> Optional[LintResult]:
+        """Inconsistent column references in GROUP BY/ORDER BY clauses."""
+        # Config type hints
+        self.group_by_and_order_by_style: str
+
+        # We only care about GROUP BY/ORDER BY clauses.
+        if context.segment.name not in {"GroupByClauseSegment", "OrderByClauseSegment"}:
+            return None
+
+        # Look at child segments to detect the presence of disallowed segments.
+        child_segment_names = [segment.name for segment in context.segment.segments]
+        disallowed_segment_names = {
+            "explicit": "numeric_literal",
+            "implicit": "ColumnReferenceSegment",
```

Comment:
`ORDER BY` and `GROUP BY` also support expressions. We should make sure the rule handles these correctly, e.g.:
```
select * from foo order by c1 * c1;
select * from foo order by power(c1, 2);
select * from foo order by 1 / c1;
```

Reply:
Would be good to add a test case for this. I was assuming it would be ignored as neither numeric nor column reference is a direct child but would be good to confirm.

Reply:
@barrywhart Sure I can account for those, I think surely an expression must contain explicit column references, as `1 / 1` or `power(1, 2)` couldn't be valid with positional references. So can simplify the logic by considering an expression as explicit?

Reply:
I think so yes (i.e. expressions are explicit).

Reply:
@barrywhart I've added the unit tests for explicit expressions and works fine for the `consistent`/`explicit` config options.

However, if a user were to set the config option as implicit (so they require all positional references) they would get unfixable lint errors if they were to use an expression as it is explicit and can't be made implicit. This isn't an issue with the code as the logic is consistent, so my thought is if they want to use expressions they shouldn't choose to force implicit group by references.

Reply:
Basically they are setting inappropriate configs for their use case if they're using expressions and should consider using `consistent` instead, so don't see this as an issue 👍